src/BulkWriteCommandBuilder.php

```diff
@@ -0,0 +1,226 @@
+<?php
+
+namespace MongoDB;
+
+use MongoDB\Codec\DocumentCodec;
+use MongoDB\Codec\Encoder;
+use MongoDB\Driver\BulkWriteCommand;
+use MongoDB\Driver\Manager;
+use MongoDB\Exception\InvalidArgumentException;
+
+use function is_array;
+use function is_bool;
+use function is_string;
+
+readonly class BulkWriteCommandBuilder
+{
+    private function __construct(
+        public BulkWriteCommand $bulkWriteCommand,
+        private Manager $manager,
+        private string $namespace,
+        private Encoder $builderEncoder,
+        private ?DocumentCodec $codec,
+    ) {
+    }
+
+    public static function createWithCollection(Collection $collection, array $options): self
+    {
+        $options += ['ordered' => true];
+
+        if (isset($options['bypassDocumentValidation']) && ! is_bool($options['bypassDocumentValidation'])) {
+            throw InvalidArgumentException::invalidType('"bypassDocumentValidation" option', $options['bypassDocumentValidation'], 'boolean');
+        }
+
+        if (isset($options['let']) && ! is_document($options['let'])) {
+            throw InvalidArgumentException::expectedDocumentType('"let" option', $options['let']);
+        }
+
+        if (! is_bool($options['ordered'])) {
+            throw InvalidArgumentException::invalidType('"ordered" option', $options['ordered'], 'boolean');
+        }
+
+        if (isset($options['verboseResults']) && ! is_bool($options['verboseResults'])) {
+            throw InvalidArgumentException::invalidType('"verboseResults" option', $options['verboseResults'], 'boolean');
+        }
+
+        return new self(
+            new BulkWriteCommand($options),
+            $collection->getManager(),
+            $collection->getNamespace(),
+            $collection->getBuilderEncoder(),
+            $collection->getCodec(),
+        );
+    }
+
+    public function withCollection(Collection $collection): self
+    {
+        /* Prohibit mixing Collections associated with different Manager
+         * objects. This is not technically necessary, since the Collection is
+         * only used to derive a namespace and encoding options; however, it
+         * may prevent a user from inadvertently mixing writes destined for
+         * different deployments. */
+        if ($this->manager !== $collection->getManager()) {
+            throw new InvalidArgumentException('$collection is associated with a different MongoDB\Driver\Manager');
+        }
+
+        return new self(
```

Comment:
Do you allow extending this class? In that case, you should use `static` here and in all return types. Otherwise the class should be final.

Reply:
Introducing it as `final` seems preferable.